{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Megabytes processed: 1.000033\n",
      "Megabytes processed: 2.000115\n",
      "Megabytes processed: 3.000093\n",
      "Megabytes processed: 4.000020\n",
      "Megabytes processed: 5.000102\n",
      "Megabytes processed: 6.000038\n",
      "Megabytes processed: 7.000046\n",
      "Megabytes processed: 8.000099\n",
      "Megabytes processed: 9.000115\n",
      "Megabytes processed: 10.000108\n",
      "Megabytes processed: 11.000114\n",
      "Megabytes processed: 12.000221\n",
      "Megabytes processed: 13.000135\n",
      "Megabytes processed: 14.000030\n",
      "Megabytes processed: 15.000022\n",
      "Megabytes processed: 16.000010\n",
      "Megabytes processed: 17.000097\n",
      "Megabytes processed: 18.000151\n",
      "Megabytes processed: 19.000050\n",
      "Megabytes processed: 20.000146\n",
      "Megabytes processed: 21.000066\n",
      "Megabytes processed: 22.000041\n",
      "Megabytes processed: 23.000149\n",
      "Megabytes processed: 24.000045\n",
      "Megabytes processed: 25.000022\n",
      "Megabytes processed: 26.000042\n",
      "Megabytes processed: 27.000049\n",
      "Megabytes processed: 28.000142\n",
      "Megabytes processed: 29.000174\n",
      "Megabytes processed: 30.000055\n",
      "Megabytes processed: 31.000116\n",
      "Megabytes processed: 32.000037\n",
      "Megabytes processed: 33.000079\n",
      "Megabytes processed: 34.000029\n",
      "Megabytes processed: 35.000134\n",
      "Megabytes processed: 36.000097\n",
      "Megabytes processed: 37.000154\n",
      "Megabytes processed: 38.000003\n",
      "Megabytes processed: 39.000092\n",
      "Megabytes processed: 40.000103\n",
      "Megabytes processed: 41.000002\n",
      "Megabytes processed: 42.000063\n",
      "Megabytes processed: 43.000045\n",
      "Megabytes processed: 44.000102\n",
      "Megabytes processed: 45.000065\n",
      "Megabytes processed: 46.000094\n",
      "Megabytes processed: 47.000090\n",
      "Megabytes processed: 48.000004\n",
      "Megabytes processed: 49.000091\n",
      "Megabytes processed: 50.000068\n",
      "Megabytes processed: 51.000131\n",
      "Megabytes processed: 52.000068\n",
      "Megabytes processed: 53.000039\n",
      "Megabytes processed: 54.000017\n",
      "Megabytes processed: 55.000084\n",
      "Megabytes processed: 56.000115\n",
      "Megabytes processed: 57.000072\n",
      "Megabytes processed: 58.000105\n",
      "Megabytes processed: 59.000011\n",
      "Megabytes processed: 60.000085\n",
      "Megabytes processed: 61.000037\n",
      "Megabytes processed: 62.000047\n",
      "Megabytes processed: 63.000035\n",
      "Megabytes processed: 64.000081\n",
      "Megabytes processed: 65.000126\n",
      "Megabytes processed: 66.000031\n",
      "Megabytes processed: 67.000015\n",
      "Megabytes processed: 68.000037\n",
      "Megabytes processed: 69.000003\n",
      "Megabytes processed: 70.000001\n",
      "Megabytes processed: 71.000018\n",
      "Megabytes processed: 72.000052\n",
      "Megabytes processed: 73.000103\n",
      "Megabytes processed: 74.000033\n",
      "Megabytes processed: 75.000040\n",
      "Megabytes processed: 76.000105\n",
      "Megabytes processed: 77.000137\n",
      "Megabytes processed: 78.000089\n",
      "Megabytes processed: 79.000012\n",
      "Megabytes processed: 80.000071\n",
      "Megabytes processed: 81.000040\n",
      "Megabytes processed: 82.000006\n",
      "Megabytes processed: 83.000002\n",
      "Megabytes processed: 84.000035\n",
      "Megabytes processed: 85.000051\n",
      "Megabytes processed: 86.000089\n",
      "Megabytes processed: 87.000029\n",
      "Megabytes processed: 88.000081\n"
     ]
    }
   ],
   "source": [
    "i=1\n",
    "i2 = 1\n",
    "#Don't worry about the i and i2. Kelsey put that in as a check to make sure the file was being processed. It has to do with the calculation below that shows how many megabytes have been processed\n",
    "gff3_file = 'Elegans.gff3'\n",
    "output_CDS = 'Elegans_CDS.gff3'\n",
    "#input file is the GFF3 of the relevant C elegans release and the output file is where the filtered info will go \n",
    "#first part of the code here prints how many megabytes are processed from the original file \n",
    "with open(gff3_file,'r') as infile, open(output_CDS,'w') as outfile:\n",
    "    while True:\n",
    "        if i/(2**20) >= i2:\n",
    "            print('Megabytes processed: %f' % (i/(2**20)))\n",
    "            i2 += 1\n",
    "        l = infile.readline()\n",
    "        if len(l) == 0:\n",
    "            break\n",
    "#After reading the file in you want to trim the white space after each row and you want to create a list of strings for the data in each column. That is what l.strip and split do\n",
    "        ltrim = l.strip()\n",
    "        lcols = ltrim.split()\n",
    "#To keep info at the top of GFF file you want to write the lines starting with #onto the output and then you want to write only those lines which have CDS in col 2 and WormBase in col 1\n",
    "        if lcols[0][0] == '#':\n",
    "            outfile.write(l)\n",
    "        elif lcols[2] == 'CDS' and lcols[1] == 'WormBase':\n",
    "            outfile.write(l)\n",
    "        i+=len(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##gff-version 3\n",
      "\n",
      "##sequence-region   I 1 15072434\n",
      "\n",
      "##sequence-region   II 1 15279421\n",
      "\n",
      "##sequence-region   III 1 13783801\n",
      "\n",
      "##sequence-region   IV 1 17493829\n",
      "\n",
      "##sequence-region   MtDNA 1 13794\n",
      "\n",
      "##sequence-region   V 1 20924180\n",
      "\n",
      "##sequence-region   X 1 17718942\n",
      "\n",
      "#!genome-build WormBase WBcel235\n",
      "\n",
      "#!genome-version WBcel235\n",
      "\n",
      "#!genome-date 2012-12\n",
      "\n",
      "#!genome-build-accession NCBI:GCA_000002985.3\n",
      "\n",
      "#!genebuild-last-updated 2019-01\n",
      "\n",
      "###\n",
      "\n",
      "###\n",
      "\n",
      "I\tWormBase\tCDS\t4221\t4358\t.\t-\t0\tID=CDS:Y74C9A.3.1;Parent=transcript:Y74C9A.3.1;protein_id=Y74C9A.3.1\n",
      "\n",
      "I\tWormBase\tCDS\t5195\t5296\t.\t-\t0\tID=CDS:Y74C9A.3.1;Parent=transcript:Y74C9A.3.1;protein_id=Y74C9A.3.1\n",
      "\n",
      "I\tWormBase\tCDS\t6037\t6327\t.\t-\t0\tID=CDS:Y74C9A.3.1;Parent=transcript:Y74C9A.3.1;protein_id=Y74C9A.3.1\n",
      "\n",
      "I\tWormBase\tCDS\t9727\t9846\t.\t-\t0\tID=CDS:Y74C9A.3.1;Parent=transcript:Y74C9A.3.1;protein_id=Y74C9A.3.1\n",
      "\n",
      "I\tWormBase\tCDS\t10095\t10148\t.\t-\t0\tID=CDS:Y74C9A.3.1;Parent=transcript:Y74C9A.3.1;protein_id=Y74C9A.3.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#This is just to show that the output is a file with CDS info from Wormbase \n",
    "gff3file = open(\"Elegans_CDS.gff3\")\n",
    "nlines = 20 \n",
    "for i in range(nlines):\n",
    "    line = gff3file.readline()\n",
    "    print(line)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
